### ISSUE TITLE
JIT invariance problem for large Python integers as arguments to function

### ISSUE THREAD

**Issue by hawkinsp** [google/jax#6684]
```
In [1]: import jax, jax.numpy as jnp

In [2]: jnp.multiply(2 ** 100, 3.)  # doesn't crash
   ...:
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
Out[2]: DeviceArray(3.8029518e+30, dtype=float32)

In [3]: jax.jit(jnp.multiply)(2 ** 100, 3.)
...
OverflowError                             Traceback (most recent call last)
<ipython-input-3-2aed3283764c> in <module>
----> 1 jax.jit(jnp.multiply)(2 ** 100, 3.)

~/p/jax/jax/_src/dtypes.py in _scalar_type_to_dtype(typ, value)
    125   if typ is int and value is not None:
    126     if value < np.iinfo(dtype).min or value > np.iinfo(dtype).max:
--> 127       raise OverflowError(f"Python int {value} too large to convert to {dtype}")
    128   return dtype
    129

OverflowError: Python int 1267650600228229401496703205376 too large to convert to int32
```

**Comment by hawkinsp** [comment]
Also note this comes from https://github.com/google/jax/pull/6165 which has a test in `api_test.py`.

**Comment by hawkinsp** [comment]
Matt's thought on this was:
basically, when we device_put values for executing a jitted computation, we should look at the corresponding aval in the jaxpr
(rather than looking just at the type of the value we're putting, and the current value of the x64 flag)

**Comment by jakevdp** [comment]
Looking at #6165 now... I wish I'd seen that earlier. I think I previously took out the specialization on int input that #6165 added back in, specifically to prevent this kind of issue

**Comment by jakevdp** [comment]
I think we have three possible approaches to this jit invariance issue:

1. revert #6165 so non-jitted code fails like jitted code
2. entirely re-design how constants are lowered during tracing so that they are converted to the type eventually used in the jaxpr, rather than to the type reflecting their intrinsic Python type
3. Accept and document this jit invariance breakage

Thoughts?

**Comment by mattjj** [comment]
I want to try 2! After all, we already want it for other reasons: to make the x64 context manager work correctly with jit. (Btw see the jax-dev chat for more context on this.)

**Comment by mattjj** [comment]
The backup plan would be 1, but that'd require changing some downstream user code. (It might be a minor change though; hard to say without investigating it.)

**Comment by mattjj** [comment]
@jakevdp do you want to pursue plan 1 now (with a global presubmit check)? Then if that works out easily, i.e. there isn't much downstream code to fix, we can unblock jit invariance right away. If it _doesn't_ work easily, at least we'll have more motivation for plan 2.

I plan to revise constant handling stuff for the x64 context manager, and I'll see if plan 2 can be folded into that easily. But this way we might not have to block on it.

WDYT?

**Comment by jakevdp** [comment]
If we're going to pursue plan 2, then in this case I think we can leave this issue open for now and reduce churn. This seems like a pretty low-priority JIT invariance case, with an already good error message and a pretty clear fix (manually casting the input to the desired type)

**Comment by jakevdp** [comment]
I think the issue has been fixed:
```python
In [1]: import jax, jax.numpy as jnp                                                                                                                                                                                                                                                                            

In [2]: jnp.multiply(2 ** 100, 3.)                                                                                                                                                                                                                                                                              
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
<ipython-input-2-b332690c3005> in <module>
----> 1 jnp.multiply(2 ** 100, 3.)

    [... skipping hidden 6 frame]

~/github/google/jax/jax/_src/dtypes.py in _scalar_type_to_dtype(typ, value)
    149   if typ is int and value is not None:
    150     if value < np.iinfo(dtype).min or value > np.iinfo(dtype).max:
--> 151       raise OverflowError(f"Python int {value} too large to convert to {dtype}")
    152   return dtype
    153 

OverflowError: Python int 1267650600228229401496703205376 too large to convert to int32
```
